commit feb8ca37cc3d83c07fd042509ef1e176cfeb2cfa

From: Trond Myklebust <Trond.Myklebust@netapp.com>

SUNRPC: Ensure that we honour autoclose before attempting to reconnect

    If the XPRT_CLOSE_WAIT flag is set, we need to ensure that we call
    xprt->ops->close() while holding xprt_lock_write() before we can
    start reconnecting.

    Signed-off-by: Trond Myklebust <Trond.Myklebust@netapp.com>
---

 net/sunrpc/xprt.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)


diff --git a/net/sunrpc/xprt.c b/net/sunrpc/xprt.c
index 06ca058..1b04e60 100644
--- a/net/sunrpc/xprt.c
+++ b/net/sunrpc/xprt.c
@@ -697,6 +697,10 @@ void xprt_connect(struct rpc_task *task)
 	}
 	if (!xprt_lock_write(xprt, task))
 		return;
+
+	if (test_and_clear_bit(XPRT_CLOSE_WAIT, &xprt->state))
+		xprt->ops->close(xprt);
+
 	if (xprt_connected(xprt))
 		xprt_release_write(xprt, task);
 	else {
